File: src/main/java/com/rh360/rh360/controller/UsersController.java
package com.rh360.rh360.controller;

import java.util.UUID;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

import com.rh360.rh360.dto.UserRequest;
import com.rh360.rh360.dto.UserResponse;
import com.rh360.rh360.entity.User;
import com.rh360.rh360.service.UsersService;
import com.rh360.rh360.util.SecurityUtil;

import jakarta.servlet.http.HttpServletRequest;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;

@RestController
@RequestMapping("/api/users")
@Tag(name = "Usuários", description = "Endpoints para gerenciamento de usuários")
public class UsersController {

    private final UsersService service;

    public UsersController(UsersService service) {
        this.service = service;
    }

    @Operation(
        summary = "Criar novo usuário",
        description = "Cria um novo usuário no sistema. Aceita JSON (application/json) ou multipart/form-data. " +
                      "Os campos 'role' e 'status' são definidos automaticamente se não fornecidos. " +
                      "O email deve ser único no sistema."
    )
    @ApiResponses(value = {
        @ApiResponse(
            responseCode = "200",
            description = "Usuário criado com sucesso",
            content = @Content(schema = @Schema(implementation = UserResponse.class))
        ),
        @ApiResponse(
            responseCode = "400",
            description = "Requisição inválida - dados do usuário incorretos ou faltando",
            content = @Content
        ),
        @ApiResponse(
            responseCode = "409",
            description = "Conflito - email já cadastrado no sistema",
            content = @Content
        ),
        @ApiResponse(
            responseCode = "500",
            description = "Erro interno do servidor",
            content = @Content
        )
    })
    @PostMapping(consumes = {"application/json"})
    public UserResponse create(@RequestBody UserRequest userRequest) {
        User user = new User();
        user.setName(userRequest.getName());
        user.setEmail(userRequest.getEmail());
        user.setPassword(userRequest.getPassword());
        user.setRole(userRequest.getRole());
        user.setStatus(userRequest.getStatus());
        User createdUser = service.create(user, null, userRequest.getPermissions());
        return new UserResponse(createdUser);
    }

    @PostMapping(consumes = {"multipart/form-data"})
    public UserResponse create(
            @RequestParam("name") String name,
            @RequestParam("email") String email,
            @RequestParam("password") String password,
            @RequestParam(value = "role", required = false) String role,
            @RequestParam(value = "status", required = false) String status,
            @RequestParam(value = "photo", required = false) MultipartFile photo) {
        User user = new User();
        user.setName(name);
        user.setEmail(email);
        user.setPassword(password);
        user.setRole(role);
        user.setStatus(status);
        User createdUser = service.create(user, photo);
        return new UserResponse(createdUser);
    }

    @Operation(
        summary = "Listar todos os usuários",
        description = "Retorna uma lista paginada com todos os usuários cadastrados. " +
                      "Parâmetros de query aceitos: " +
                      "'page' (número da página, começa em 0, padrão: 0), " +
                      "'size' (tamanho da página, padrão: 20), " +
                      "'sort' (campo de ordenação no formato 'campo,direção', exemplo: 'name,asc' ou 'createdAt,desc').",
        security = @SecurityRequirement(name = "Bearer Authentication")
    )
    @ApiResponses(value = {
        @ApiResponse(
            responseCode = "200",
            description = "Lista de usuários retornada com sucesso",
            content = @Content(schema = @Schema(implementation = Page.class))
        ),
        @ApiResponse(
            responseCode = "401",
            description = "Não autenticado - token JWT ausente ou inválido",
            content = @Content
        ),
        @ApiResponse(
            responseCode = "400",
            description = "Parâmetros de paginação inválidos",
            content = @Content
        ),
        @ApiResponse(
            responseCode = "500",
            description = "Erro interno do servidor",
            content = @Content
        )
    })
    @GetMapping
    public Page<UserResponse> findAll(Pageable pageable) {
        return service.findAll(pageable);
    }

    @Operation(
        summary = "Buscar usuário por ID",
        description = "Retorna os dados de um usuário específico pelo seu ID",
        security = @SecurityRequirement(name = "Bearer Authentication")
    )
    @ApiResponses(value = {
        @ApiResponse(
            responseCode = "200",
            description = "Usuário encontrado",
            content = @Content(schema = @Schema(implementation = UserResponse.class))
        ),
        @ApiResponse(
            responseCode = "401",
            description = "Não autenticado - token JWT ausente ou inválido",
            content = @Content
        ),
        @ApiResponse(
            responseCode = "404",
            description = "Usuário não encontrado",
            content = @Content
        ),
        @ApiResponse(
            responseCode = "400",
            description = "ID inválido",
            content = @Content
        ),
        @ApiResponse(
            responseCode = "500",
            description = "Erro interno do servidor",
            content = @Content
        )
    })
    @GetMapping("/{id}")
    public UserResponse findById(@PathVariable UUID id) {
        User user = service.findById(id);
        if (user == null) {
            throw new RuntimeException("Usuário não encontrado");
        }
        return new UserResponse(user);
    }

    @Operation(
        summary = "Obter usuário atual",
        description = "Retorna os dados do usuário autenticado baseado no token JWT",
        security = @SecurityRequirement(name = "Bearer Authentication")
    )
    @ApiResponses(value = {
        @ApiResponse(
            responseCode = "200",
            description = "Dados do usuário atual retornados com sucesso",
            content = @Content(schema = @Schema(implementation = UserResponse.class))
        ),
        @ApiResponse(
            responseCode = "401",
            description = "Não autenticado - token JWT ausente, inválido ou expirado",
            content = @Content
        ),
        @ApiResponse(
            responseCode = "404",
            description = "Usuário não encontrado",
            content = @Content
        ),
        @ApiResponse(
            responseCode = "500",
            description = "Erro interno do servidor",
            content = @Content
        )
    })
    @GetMapping("/me")
    public UserResponse getCurrentUser(HttpServletRequest request) {
        UUID userId = SecurityUtil.getUserId(request);
        if (userId == null) {
            throw new RuntimeException("Usuário não autenticado");
        }
        User user = service.findById(userId);
        if (user == null) {
            throw new RuntimeException("Usuário não encontrado");
        }
        return new UserResponse(user);
    }

    @Operation(
        summary = "Atualizar usuário",
        description = "Atualiza os dados de um usuário específico. Aceita JSON (application/json) ou multipart/form-data. " +
                      "O email deve ser único no sistema.",
        security = @SecurityRequirement(name = "Bearer Authentication")
    )
    @ApiResponses(value = {
        @ApiResponse(
            responseCode = "200",
            description = "Usuário atualizado com sucesso",
            content = @Content(schema = @Schema(implementation = UserResponse.class))
        ),
        @ApiResponse(
            responseCode = "401",
            description = "Não autenticado - token JWT ausente ou inválido",
            content = @Content
        ),
        @ApiResponse(
            responseCode = "404",
            description = "Usuário não encontrado",
            content = @Content
        ),
        @ApiResponse(
            responseCode = "400",
            description = "Dados inválidos - requisição malformada",
            content = @Content
        ),
        @ApiResponse(
            responseCode = "409",
            description = "Conflito - email já cadastrado no sistema",
            content = @Content
        ),
        @ApiResponse(
            responseCode = "500",
            description = "Erro interno do servidor",
            content = @Content
        )
    })
    @PutMapping(value = "/{id}", consumes = {"application/json"})
    public UserResponse update(@PathVariable UUID id, @RequestBody UserRequest userRequest) {
        User user = new User();
        user.setName(userRequest.getName());
        user.setEmail(userRequest.getEmail());
        user.setPassword(userRequest.getPassword());
        user.setRole(userRequest.getRole());
        user.setStatus(userRequest.getStatus());
        User updatedUser = service.update(id, user, null, userRequest.getPermissions());
        return new UserResponse(updatedUser);
    }

    @PutMapping(value = "/{id}", consumes = {"multipart/form-data"})
    public UserResponse update(
            @PathVariable UUID id,
            @RequestParam("name") String name,
            @RequestParam("email") String email,
            @RequestParam(value = "password", required = false) String password,
            @RequestParam(value = "role", required = false) String role,
            @RequestParam(value = "status", required = false) String status,
            @RequestParam(value = "photo", required = false) MultipartFile photo) {
        User user = new User();
        user.setName(name);
        user.setEmail(email);
        user.setPassword(password);
        user.setRole(role);
        user.setStatus(status);
        User updatedUser = service.update(id, user, photo);
        return new UserResponse(updatedUser);
    }

    @Operation(
        summary = "Deletar usuário",
        description = "Realiza soft delete de um usuário específico (marca como 'deleted')",
        security = @SecurityRequirement(name = "Bearer Authentication")
    )
    @ApiResponses(value = {
        @ApiResponse(
            responseCode = "200",
            description = "Usuário deletado com sucesso"
        ),
        @ApiResponse(
            responseCode = "401",
            description = "Não autenticado - token JWT ausente ou inválido",
            content = @Content
        ),
        @ApiResponse(
            responseCode = "404",
            description = "Usuário não encontrado",
            content = @Content
        ),
        @ApiResponse(
            responseCode = "400",
            description = "ID inválido",
            content = @Content
        ),
        @ApiResponse(
            responseCode = "500",
            description = "Erro interno do servidor",
            content = @Content
        )
    })
    @DeleteMapping("/{id}")
    public void delete(@PathVariable UUID id) {
        service.delete(id);
    }
}

File: src/users/users.controller.ts
import {
  Controller,
  Get,
  Post,
  Put,
  Delete,
  Body,
  Param,
  Query,
  Headers,
  UseGuards,
  UseInterceptors,
  UploadedFile,
  HttpStatus,
  HttpException,
} from '@nestjs/common';
import { FileInterceptor } from '@nestjs/platform-express';
import {
  ApiTags,
  ApiOperation,
  ApiResponse,
  ApiBearerAuth,
  ApiConsumes,
  ApiBody,
} from '@nestjs/swagger';
import { JwtAuthGuard } from '../auth/guards/jwt-auth.guard';
import { UserId } from '../auth/decorators/user.decorator';
import { TokenService } from '../token/token.service';
import { UsersService } from './users.service';
import { UserRequestDto } from './dto/user-request.dto';
import { UserUpdateRequestDto } from './dto/user-update-request.dto';
import { UserResponseDto } from './dto/user-response.dto';

@ApiTags('Usuários')
@Controller('api/users')
export class UsersController {
  constructor(
    private readonly usersService: UsersService,
    private readonly tokenService: TokenService,
  ) {}

  @ApiOperation({
    summary: 'Criar novo usuário',
    description:
      'Aceita JSON ou FormData (multipart/form-data). Foto opcional no campo "photo".',
  })
  @ApiConsumes('application/json', 'multipart/form-data')
  @ApiBody({
    schema: {
      type: 'object',
      properties: {
        name: { type: 'string' },
        email: { type: 'string' },
        password: { type: 'string' },
        role: { type: 'string' },
        status: { type: 'string' },
        permissions: { type: 'string', description: 'JSON array (FormData)' },
        photo: { type: 'string', format: 'binary' },
      },
      required: ['name', 'email'],
    },
  })
  @ApiResponse({
    status: 200,
    description: 'Usuário criado com sucesso',
    type: UserResponseDto,
  })
  @Post()
  @UseInterceptors(FileInterceptor('photo'))
  async create(
    @Body() userDto: UserRequestDto,
    @UploadedFile() photo?: Express.Multer.File,
  ): Promise<UserResponseDto> {
    try {
      return await this.usersService.create(userDto, photo);
    } catch (error) {
      throw new HttpException(
        { error: error.message },
        error.status || HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @ApiOperation({ summary: 'Listar todos os usuários' })
  @ApiBearerAuth()
  @ApiResponse({
    status: 200,
    description: 'Lista de usuários retornada com sucesso',
  })
  @UseGuards(JwtAuthGuard)
  @Get()
  async findAll(
    @Query('page') page: string = '0',
    @Query('size') size: string = '20',
    @Query('search') search?: string,
  ) {
    return await this.usersService.findAll(
      parseInt(page),
      parseInt(size),
      search,
    );
  }
  @ApiOperation({ summary: 'Obter usuário atual' })
  @ApiBearerAuth()
  @ApiResponse({
    status: 200,
    description: 'Dados do usuário atual retornados com sucesso',
    type: UserResponseDto,
  })
  
  @UseGuards(JwtAuthGuard)
  @Get('me')
  async getCurrentUser(
    @Headers('authorization') authHeader: string,
  ): Promise<UserResponseDto> {
    if (!authHeader || !authHeader.startsWith('Bearer ')) {
      throw new HttpException(
        { error: 'Token não fornecido ou formato inválido' },
        HttpStatus.UNAUTHORIZED,
      );
    }

    const token = authHeader.substring(7).trim();
    const userId = this.tokenService.extractUserId(token);
    return await this.usersService.findById(userId);
  }

  @ApiOperation({ summary: 'Buscar usuário por ID' })
  @ApiBearerAuth()
  @ApiResponse({
    status: 200,
    description: 'Usuário encontrado',
    type: UserResponseDto,
  })
  @UseGuards(JwtAuthGuard)
  @Get(':id')
  async findById(@Param('id') id: string): Promise<UserResponseDto> {
    return await this.usersService.findById(id);
  }

  @ApiOperation({ summary: 'Atualizar usuário com foto (FormData)' })
  @ApiBearerAuth()
  @ApiConsumes('multipart/form-data')
  @ApiBody({
    schema: {
      type: 'object',
      properties: {
        name: { type: 'string' },
        email: { type: 'string' },
        password: { type: 'string' },
        role: { type: 'string' },
        status: { type: 'string' },
        permissions: { type: 'string', description: 'JSON array de permissões' },
        photo: { type: 'string', format: 'binary' },
      },
    },
  })
  @ApiResponse({
    status: 200,
    description: 'Usuário atualizado com sucesso',
    type: UserResponseDto,
  })
  @UseGuards(JwtAuthGuard)
  @Put(':id/with-photo')
  @UseInterceptors(FileInterceptor('photo'))
  async updateWithPhoto(
    @Param('id') id: string,
    @Body() userDto: UserUpdateRequestDto,
    @UploadedFile() photo?: Express.Multer.File,
  ): Promise<UserResponseDto> {
    try {
      return await this.usersService.update(id, userDto, photo);
    } catch (error) {
      throw new HttpException(
        { error: error.message },
        error.status || HttpStatus.BAD_REQUEST,
      );
    }
  }

  @ApiOperation({ summary: 'Atualizar usuário (JSON ou FormData)' })
  @ApiBearerAuth()
  @ApiResponse({
    status: 200,
    description: 'Usuário atualizado com sucesso',
    type: UserResponseDto,
  })
  @UseGuards(JwtAuthGuard)
  @Put(':id')
  async update(
    @Param('id') id: string,
    @Body() userDto: UserUpdateRequestDto,
  ): Promise<UserResponseDto> {
    try {
      return await this.usersService.update(id, userDto);
    } catch (error) {
      throw new HttpException(
        { error: error.message },
        error.status || HttpStatus.BAD_REQUEST,
      );
    }
  }

  @ApiOperation({ summary: 'Deletar usuário' })
  @ApiBearerAuth()
  @ApiResponse({ status: 200, description: 'Usuário deletado com sucesso' })
  @UseGuards(JwtAuthGuard)
  @Delete(':id')
  async delete(@Param('id') id: string): Promise<void> {
    try {
      await this.usersService.delete(id);
    } catch (error) {
      throw new HttpException(
        { error: error.message },
        error.status || HttpStatus.BAD_REQUEST,
      );
    }
  }
}
